tests/qos/test_tunnel_qos_remap.py

```diff
@@ -481,6 +485,191 @@ def test_pfc_pause_extra_lossless_active(ptfhost, fanouthosts, rand_selected_dut
+        testutils.send(ptfadapter, src_port, pkt, 1)
+        # Get the actual egress port
+        result = testutils.verify_packet_any_port(ptfadapter, exp_pkt, dst_ports)
+        actual_port = dst_ports[result[0]]
+        # Get the port name from mgfacts
+        for port_name, idx in mg_facts['minigraph_ptf_indices'].items():
+            if idx == actual_port:
+                actual_port_name = port_name
+                break
+        pytest_assert(actual_port_name)
+        # Clear queue counter
+        get_queue_counter(rand_selected_dut, actual_port_name, queue, True)
+        # Clear queue watermark
+        get_queue_watermark(rand_selected_dut, actual_port_name, wmk_stat_queue, True)
+        # Send initial uncongested traffic to increase the watermark
+        testutils.send(ptfadapter, src_port, pkt, 1000)
+        # Record the base queue count
+        base_queue_count = get_queue_counter(rand_selected_dut, actual_port_name, queue, False)
+        logger.info("Base queue count on {}|{} is {}".format(actual_port_name, queue, base_queue_count))
+        assert base_queue_count >= 1000, \
+            "Incorrect number of packets arrived on {}|{}".format(actual_port_name, queue)
+        # Record watermark when under no congestion
+        base_queue_wmk = get_queue_watermark(rand_selected_dut, actual_port_name, wmk_stat_queue)
+        logger.info("Base queue watermark on {}|{} is {}".format(
+            actual_port_name, wmk_stat_queue, base_queue_wmk))
+        peer_info = leaf_fanout_peer_info(rand_selected_dut, conn_graph_facts, mg_facts, actual_port)
+        storm_handler = PFCStorm(rand_selected_dut, fanout_graph_facts, fanouthosts,
+                                 pfc_queue_idx=prio,
+                                 pfc_frames_number=PFC_PKT_COUNT,
+                                 peer_info=peer_info)
+        # Start PFC storm from leaf fanout switch
+        start_pfc_storm(storm_handler, peer_info, prio)
+        try:
+            # Send congested traffic
+            testutils.send(ptfadapter, src_port, pkt, 10000)
+        finally:
+            stop_pfc_storm(storm_handler)
+        # Clear out packets for futher verification
+        testutils.count_matched_packets_all_ports(ptfadapter, exp_pkt, dst_ports, timeout=0.5)
+        # Verify queue count increased
+        base_queue_count = get_queue_counter(rand_selected_dut, actual_port_name, queue, False)
+        logger.info("Queue count after data on {}|{} is {}".format(
+            actual_port_name, queue, base_queue_count))
+        # Record new watermark after congestion and clear
+        queue_wmk = get_queue_watermark(rand_selected_dut, actual_port_name, wmk_stat_queue, True)
+        logger.info("Congested queue watermark on {}|{} is {}".format(
+            actual_port_name, wmk_stat_queue, queue_wmk))
+        assert queue_wmk > base_queue_wmk, \
+            "Failed to detect congestion due to PFC pause, failed check {} > {}".format(
```

Comment:
Can we add some threshold for checking the queue watermark?

Reply:
Do you mean something like `queue_wmk > base_queue_wmk + x`, where `x` is some percentage of the traffic sent that is expected to be enqueued due to pfc_gen.py congestion?

Reply:
Yes, I think that makes the test more reliable.

Reply:
Added a threshold expecting a 1% increase. 
Couple notes on this:
- The original test does send 1000 baseline packets to ensure the watermark is at a reasonable uncongested baseline. Packets are sent so slowly that there's very little likelihood of causing any passive watermark increase (e.g. from near-line-rate forwarding). The 10000 packet data in the PFC XOFF frame congested state should then only need a non-zero increase beyond this baseline. 
- For the thresholded change, since the data load is extremely slow and a noisy pfc_gen.py congestion FPS will allow frames to egress at extremely frequent intervals, only a very small percentage of the traffic can be expected to have been congested at once. In my experiments, we see an increase of less than 10%, so I set the margin to 1%. Technically any non-zero hardcoded number here will be arbitrary, not sure about possible heuristics.

Reply:
Thanks @rbpittman . That makes sense.